File: box-and-box/examples/supervise.mjs
// examples/supervise.mjs — the temporal layer over an agent's trajectory.
// States carry the invariant Value's fields, so temporal atoms are predicates over the
// real substrate: G(β≥0.8) = "confidence never drops"; F(done) = "the task completes".
import { atom, always, eventually, show } from '../temporal.mjs';
import { TemporalSpec, supervise, residualOf, guard } from '../supervise.mjs';

const betaOK = atom('β≥0.8', (s) => s.beta >= 0.8);
const done = atom('done', (s) => s.done === true);

const specs = [
  TemporalSpec({ id: 'confidence-floor', formula: always(betaOK), kind: 'safety' }),
  TemporalSpec({ id: 'reach-goal', formula: eventually(done), kind: 'liveness', ctd: 'escalate-replan' })
];

const report = (title, τ) => {
  const r = supervise(τ, specs);
  console.log('\n' + title + '\n  ' + '─'.repeat(70));
  console.log('  trajectory: ' + τ.map((s) => `[β${s.beta}${s.done ? ',done' : ''}]`).join(' → '));
  r.reports.forEach((x) => {
    const mark = x.verdict === 'satisfied' ? '✓' : '✗';
    let line = `  ${mark} ${x.kind.padEnd(8)} ${x.id.padEnd(18)} ${show({ t: 'atom', name: x.formula })}  → ${x.verdict}`;
    if (x.violatedAt != null) line += `  (witness @ step ${x.violatedAt})`;
    if (x.escalation) line += `  → escalate: ${x.escalation}`;
    console.log(line);
    console.log('     online: ' + x.online.join(' · '));
  });
  console.log('  ' + r.note);
};

console.log('\nTemporal supervision · safety shield + liveness obligation');
report('1 · a good run (safe, reaches the goal)',
  [{ beta: 0.95, done: false }, { beta: 0.90, done: false }, { beta: 0.88, done: true }]);
report('2 · a bad run (confidence dips, goal never reached)',
  [{ beta: 0.90, done: false }, { beta: 0.50, done: false }, { beta: 0.85, done: false }]);

// the shield, one step ahead
console.log('\n3 · the safety shield (one-step lookahead)\n  ' + '─'.repeat(70));
const history = [{ beta: 0.95, done: false }, { beta: 0.90, done: false }];
const residual = residualOf(always(betaOK), history);
console.log(`  after ${history.length} safe steps, residual = ${show(residual)}`);
console.log(`  guard( next = [β0.50] ) → ${guard(residual, { beta: 0.50 })}   ${guard(residual, { beta: 0.50 }) ? '⇒ PRUNE this action' : ''}`);
console.log(`  guard( next = [β0.95] ) → ${guard(residual, { beta: 0.95 })}   ${guard(residual, { beta: 0.95 }) ? '' : '⇒ allow'}`);
console.log('\n  Safety is enforced before the step is taken; liveness can only be judged at the horizon.\n');

File: box-and-box/ports/rust/src/supervise.rs
// supervise.rs — trajectory supervision (faithful port of supervise.mjs). Laws TB1–TB3.

use crate::temporal::{monitor, progress, Formula, State};

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    Safety,
    Liveness,
}

#[derive(Clone)]
pub struct TemporalSpec {
    pub id: String,
    pub formula: Formula,
    pub kind: Kind,
    pub ctd: Option<String>,
}

impl TemporalSpec {
    pub fn new(id: &str, formula: Formula, kind: Kind, ctd: Option<String>) -> TemporalSpec {
        TemporalSpec {
            id: id.to_string(),
            formula,
            kind,
            ctd,
        }
    }
}

fn first_vio(online: &[String]) -> Option<usize> {
    online.iter().position(|s| s == "vio")
}

pub struct Report {
    pub id: String,
    pub kind: Kind,
    pub verdict: String,
    pub online: Vec<String>,
    pub violated_at: Option<usize>,
    pub escalation: Option<String>,
    pub reason: Option<String>,
}

pub struct SpecEsc {
    pub id: String,
    pub repair: String,
}

pub struct SuperviseResult {
    pub reports: Vec<Report>,
    pub safe: bool,
    pub escalation: Option<Vec<SpecEsc>>,
}

pub fn supervise(trajectory: &[State], specs: &[TemporalSpec]) -> SuperviseResult {
    let reports: Vec<Report> = specs
        .iter()
        .map(|spec| {
            let m = monitor(&spec.formula, trajectory);
            let violated_at = if spec.kind == Kind::Safety {
                first_vio(&m.online)
            } else {
                None
            };
            let (escalation, reason) =
                if spec.kind == Kind::Liveness && m.verdict == "violated" {
                    (
                        Some(spec.ctd.clone().unwrap_or_else(|| "escalate-to-human".into())),
                        Some(format!(
                            "liveness obligation unmet within horizon ({} steps)",
                            trajectory.len()
                        )),
                    )
                } else {
                    (None, None)
                };
            Report {
                id: spec.id.clone(),
                kind: spec.kind,
                verdict: m.verdict,
                online: m.online,
                violated_at,
                escalation,
                reason,
            }
        })
        .collect();

    let safe = !reports
        .iter()
        .any(|r| r.kind == Kind::Safety && r.verdict == "violated");
    let liveness_unmet: Vec<&Report> = reports
        .iter()
        .filter(|r| r.kind == Kind::Liveness && r.verdict == "violated")
        .collect();

    let escalation = if !liveness_unmet.is_empty() {
        Some(
            liveness_unmet
                .iter()
                .map(|r| SpecEsc {
                    id: r.id.clone(),
                    repair: r.escalation.clone().unwrap_or_default(),
                })
                .collect(),
        )
    } else {
        None
    };

    SuperviseResult {
        reports,
        safe,
        escalation,
    }
}

// the one-step shield
pub fn residual_of(formula: &Formula, history: &[State]) -> Formula {
    history.iter().fold(formula.clone(), |f, s| progress(&f, s))
}

pub fn guard(residual: &Formula, next_state: &State) -> bool {
    matches!(progress(residual, next_state), Formula::False)
}
